REGRESSION(290212@main): Sometimes web content never becomes accessible because the isolated tree never gets built

twilco opened the pull request:
#### b9f8037aea654c6dd7ca25330fe9e4522e99a872
<pre>
REGRESSION(290212@main): Sometimes web content never becomes accessible because the isolated tree never gets built
<a href="https://bugs.webkit.org/show_bug.cgi?id=290194">https://bugs.webkit.org/show_bug.cgi?id=290194</a>
<a href="https://rdar.apple.com/147592259">rdar://147592259</a>

Reviewed by Chris Fleizach.

In <a href="https://commits.webkit.org/290212@main">https://commits.webkit.org/290212@main</a>, we made `-[WKAccessibilityWebPageObjectBase focusedLocalFrame]` return
nullptr if off the main-thread to avoid unsafely accessing the `WebPage` (which can only be used on the main-thread).
This was the right thing to do, but had unexpected fallout, as it meant that when `-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]`
was called with a nullptr frame, we never actually called `AXObjectCache::rootObjectForFrame`, which would kick off
creation of the isolated tree.

With this commit, -[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:] now will try to re-retrieve a
frame after dispatching to the main-thread if the one we had was nullptr for any reason.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::buildAccessibilityTreeIfNeeded):
Drive-by fix to call getOrCreateIsolatedTree() instead of isolatedTreeRootObject(), as the former better describes
the intent of behavior we want to achieve in this function. There is no change in actual behavior here, and unrelated
to the bug described above.
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]):

Canonical link: <a href="https://commits.webkit.org/292613@main">https://commits.webkit.org/292613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83b47a5490b2f3ca7dca767c87c4a8be5fdadd36

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15904 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5874 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101357 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46809 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98335 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16200 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24337 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73390 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30618 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99293 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86978 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53727 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4744 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46136 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4842 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103385 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23357 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [❌ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23608 "Failed to checkout and rebase branch from PR 42823") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83002 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81807 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20593 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3878 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16731 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23320 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28475 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22979 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26459 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24720 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1075,7 +1075,7 @@ void AXObjectCache::setIsolatedTreeRoot(AXCoreObject* root)
 #endif
 
-AXCoreObject* AXObjectCache::rootObjectForFrame(LocalFrame& frame)
+AXCoreObject* AXObjectCache::rootObjectForFrame(LocalFrame* frame)
```
The name rootObjectForFrame really doesn't make sense now if the frame is optional.

What would you name it if this was a brand-new function?

How about something like: getOrCreateRootObject(LocalFrame* optionalLocalFrame)

I almost wonder if it'd be more clear to split it. Have AXObjectCache::rootObjectIfExists(), and then the codepaths that do have a LocalFrame could instead call AXObjectCache::buildFromFrame().

Also: AXObjectCache is owned by a Page and associated with a Document. Can't it get the frame from its page or document? Does it actually need LocalFrame to be passed in?

twilco replied:
> The name rootObjectForFrame really doesn't make sense now if the frame is optional.

I walked back this part of the change in the latest commit to avoid introducing this naming inconsistency. It wasn't necessary to fix the bug, and when we get around to better supporting site isolation, we probably do actually want to require a frame here.

> Also: AXObjectCache is owned by a Page and associated with a Document. Can't it get the frame from its page or document? Does it actually need LocalFrame to be passed in?

Good question. Today, we probably could get the frame from either source. But with site isolation, I don't think we can, since a page can have multiple local frames. Not 100% sure about that though, we'll need to spend some time sorting out how all of these functions need to change with site isolation.

minorninth replied:
Do we actually need a method that returns the root object for any frame, or just for the main frame? Right now it looks like it only returns the main frame and that seems to work fine. I suspect that's all we'll need for site isolation too, but let's discuss.

twilco replied:
Good question. I think we might need to do that based on what Josh has told me, but let's review some of the [site isolation documentation](https://docs.webkit.org/Deep%20Dive/SiteIsolation.html#main-and-root-frames) together and create a plan.

minorninth commented on Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
```diff
@@ -899,7 +899,7 @@ void WebFrame::setAccessibleName(const AtomString& accessibleName)
         return;
     
-    RefPtr rootObject = document->axObjectCache()->rootObjectForFrame(*localFrame);
+    RefPtr rootObject = document->axObjectCache()->rootObjectForFrame(localFrame.get());
```
Likewise, this seems slightly convoluted to me, because if we have a document, then we MUST have a frame. But the axObjectCache has access to that same document, so it could just get the frame from there?

I think a lot of this stuff evolved and made sense when it was first added, and cleaning it up may make fixing bugs like this easier.

twilco replied:
>  because if we have a document, then we MUST have a frame. But the axObjectCache has access to that same document, so it could just get the frame from there?

You're totally right. Walked this part of the change back — let's re-evaluate when we continue our site isolation effort.

minorninth reviewed (commented):
I left some comments, but the fix looks good and fine with merging this and then cleaning up in a follow-up, especially if you think I'm wrong about not even needing to pass in a frame.

twilco committed:
REGRESSION(290212@main): Sometimes web content never becomes accessible because the isolated tree never gets built
https://bugs.webkit.org/show_bug.cgi?id=290194
rdar://147592259

Reviewed by Chris Fleizach.

In https://commits.webkit.org/290212@main, we made `-[WKAccessibilityWebPageObjectBase focusedLocalFrame]` return
nullptr if off the main-thread to avoid unsafely accessing the `WebPage` (which can only be used on the main-thread).
This was the right thing to do, but had unexpected fallout, as it meant that when `-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]`
was called with a nullptr frame, we never actually called `AXObjectCache::rootObjectForFrame`, which would kick off
creation of the isolated tree.

With this commit, -[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:] now will try to re-retrieve a
frame after dispatching to the main-thread if the one we had was nullptr for any reason.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::buildAccessibilityTreeIfNeeded):
Drive-by fix to call getOrCreateIsolatedTree() instead of isolatedTreeRootObject(), as the former better describes
the intent of behavior we want to achieve in this function. There is no change in actual behavior here, and unrelated
to the bug described above.
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]):

Canonical link: https://commits.webkit.org/292613@main
